server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class MapperMergeContext {
```

Comment:
While introducing this new context may be a good idea (it is weird to use MapperBuilderContext when merging to some degree), I find that it requires a lot of mechanical changes that end up being a distraction when reviewing the core changes in this PR. Is it an option to consider making that as a follow-up perhaps? Do you consider using a separate context a requirement for this PR to go in? The point is that the builder context has stuff that is not needed to merge, and you are adding stuff that is only needed for merging? If we think it's needed, we could also make that change as a pre-requisite, it should be easier to review on its own.

Reply:
> Is it an option to consider making that as a follow-up perhaps?

I think that may lead to even more mechanical changes as we'd need to add `maxFieldsToAddDuringMerge` to `MapperBuilderContext` and then move it to `MapperBuilderContext` again.

But I could probably introduce the `MapperMergeContext` in a separate PR, as a prerequisite of this PR. On the other hand, this PR isn't exactly huge, even if it contains some mechanical changes.

But if you prefer so, I'll go ahead and do it. Let me know.

Reply:
The PR isn't huge, I agree, but not small either. The problem is when we mix mechanical changes with core changes that are delicate. I am afraid of missing something, especially as we have many merge methods and it's difficult to track which are used where. I'd be happy to review adding this MapperMergeContext separately if that's not too much work on your end.

Reply:
np, I'm working on it!

Reply:
Here's the PR: https://github.com/elastic/elasticsearch/pull/104512
Once that is merged, I'll update this PR which should drastically reduce the number of mechanical changes, exposing just the changes to the mapping merge process.

Reply:
++ thanks